File: tests/test_collections.py
"""
Test collection organization functionality.
"""
import pytest
from powerfulcases import load, cases, collections


class TestCollections:
    """Test collection listing and organization."""

    def test_collections_returns_list(self):
        """collections() returns a list of collection names."""
        colls = collections()
        assert isinstance(colls, list)
        assert len(colls) > 0

    def test_collections_includes_expected(self):
        """collections() includes expected collection names."""
        colls = collections()
        expected = ['ieee-transmission', 'ieee-distribution', 'synthetic', 'matpower', 'test']
        for coll in expected:
            assert coll in colls, f"Expected collection '{coll}' not found"

    def test_collections_sorted(self):
        """collections() returns sorted list."""
        colls = collections()
        assert colls == sorted(colls)


class TestCasesFiltering:
    """Test cases() function with filtering."""

    def test_cases_no_filter_returns_all(self):
        """cases() without filter returns all cases."""
        all_cases = cases()
        assert isinstance(all_cases, list)
        assert len(all_cases) == 88  # Total expected cases

    def test_cases_filter_by_collection(self):
        """cases(collection=...) filters correctly."""
        trans_cases = cases(collection='ieee-transmission')
        assert len(trans_cases) == 8
        expected = ['ieee14', 'ieee30', 'ieee39', 'ieee57', 'ieee118', 'ieee300', 'ieee24_rts', 'npcc']
        for case in expected:
            assert case in trans_cases

    def test_cases_filter_ieee_distribution(self):
        """IEEE distribution collection (currently empty, cases not yet moved)."""
        dist_cases = cases(collection='ieee-distribution')
        # NOTE: OpenDSS cases are still in flat structure, not moved to collection yet
        assert isinstance(dist_cases, list)

    def test_cases_filter_synthetic(self):
        """Synthetic collection has expected cases."""
        synth_cases = cases(collection='synthetic')
        assert len(synth_cases) == 10
        assert 'ACTIVSg2000' in synth_cases

    def test_cases_filter_nonexistent_collection(self):
        """Filtering by nonexistent collection returns empty list."""
        result = cases(collection='nonexistent')
        assert result == []

    def test_cases_sorted(self):
        """cases() returns sorted list."""
        all_cases = cases()
        assert all_cases == sorted(all_cases)


class TestLoadByName:
    """Test loading cases by name (searches collections)."""

    def test_load_searches_collections(self):
        """load() finds cases in collections without specifying collection."""
        case = load('ieee14')
        assert case.name == 'ieee14'
        assert case.collection == 'ieee-transmission'

    def test_load_with_collection_path(self):
        """load('collection/case') works."""
        case = load('ieee-transmission/ieee14')
        assert case.name == 'ieee14'
        assert case.collection == 'ieee-transmission'

    def test_load_unknown_case_raises(self):
        """load() raises ValueError for unknown case."""
        with pytest.raises(ValueError, match="Unknown case"):
            load('nonexistent_case_xyz')


class TestCaseBundleProperties:
    """Test CaseBundle collection and tags properties."""

    def test_collection_property(self):
        """CaseBundle.collection returns collection name."""
        case = load('ieee14')
        assert case.collection == 'ieee-transmission'

    def test_tags_property(self):
        """CaseBundle.tags returns list (empty if not defined)."""
        case = load('ieee14')
        assert isinstance(case.tags, list)
        # Tags are empty until we add them to manifests

    def test_collection_inferred_from_directory(self):
        """Collection is inferred from directory structure."""
        case = load('ieee118')
        assert case.collection == 'ieee-transmission'


class TestBackwardCompatibility:
    """Test that existing functionality still works."""

    def test_load_by_name_still_works(self):
        """Legacy flat name loading still works."""
        case = load('ieee14')
        assert case.name == 'ieee14'

    def test_cases_returns_all_cases(self):
        """cases() returns all 88 cases."""
        all_cases = cases()
        assert len(all_cases) == 88
        assert 'ieee14' in all_cases
        assert 'ACTIVSg2000' in all_cases
        assert 'case118zh' in all_cases

File: test/test_collections.jl
# Test collection organization functionality for PowerfulCases.

using Test
using PowerfulCases

@testset "Collections" begin
    @testset "collections() function" begin
        @test collections() isa Vector{String}
        @test length(collections()) > 0

        # Check expected collections exist
        colls = collections()
        @test "ieee-transmission" in colls
        @test "synthetic" in colls
        @test "matpower" in colls
        @test "test" in colls

        # Check sorted
        @test colls == sort(colls)
    end

    @testset "cases() filtering" begin
        # No filter returns all cases
        all_cases = cases()
        @test all_cases isa Vector{String}
        @test length(all_cases) == 88

        # Filter by collection
        trans_cases = cases(collection="ieee-transmission")
        @test length(trans_cases) == 8
        @test "ieee14" in trans_cases
        @test "ieee39" in trans_cases
        @test "ieee118" in trans_cases

        # Synthetic collection
        synth_cases = cases(collection="synthetic")
        @test length(synth_cases) == 10
        @test "ACTIVSg2000" in synth_cases

        # Nonexistent collection returns empty
        @test cases(collection="nonexistent") == []

        # Cases are sorted
        @test all_cases == sort(all_cases)
    end

    @testset "load() searches collections" begin
        # Load by name (searches collections)
        case = load("ieee14")
        @test case.name == "ieee14"
        @test case.collection == "ieee-transmission"

        # Load with collection/case path
        case2 = load("ieee-transmission/ieee14")
        @test case2.name == "ieee14"
        @test case2.collection == "ieee-transmission"

        # Unknown case raises error
        @test_throws ErrorException load("nonexistent_case_xyz")
    end

    @testset "CaseBundle properties" begin
        case = load("ieee14")

        # Collection property
        @test case.collection == "ieee-transmission"

        # Tags property
        @test case.tags isa Vector{String}
        # Tags are empty until we add them to manifests

        # Collection inferred from directory
        case2 = load("ieee118")
        @test case2.collection == "ieee-transmission"
    end

    @testset "Backward compatibility" begin
        # Load by name still works
        case = load("ieee14")
        @test case.name == "ieee14"

        # cases() returns all cases
        all_cases = cases()
        @test length(all_cases) == 88
        @test "ieee14" in all_cases
        @test "ACTIVSg2000" in all_cases
        @test "case118zh" in all_cases
    end
end
